servers/dsub/jobs/controllers/utils/generator.py

```diff
@@ -0,0 +1,164 @@
+        task_ids=dstat_params.get('task_ids'),
+        create_time_min=create_time_min,
+        create_time_max=create_time_max,
+        job_names=dstat_params.get('job_names'),
+        labels=dstat_params.get('labels')))
+
+    last_create_time = None
+    job_buffer = []
+    for j in jobs:
+        job = _query_jobs_result(j, proj_id)
+        # Filter pending vs. running jobs since dstat does not have
+        # a corresponding status (both RUNNING)
+        if query.statuses and job.status not in query.statuses:
+            continue
+        if query.start and (not job.start or job.start < query.start):
+            continue
+        if query.end and (not job.end or job.end > query.end):
+            continue
+
+        # If this job is from the last page, skip it and continue generating
+        if create_time_max and job.submission == create_time_max:
+            if offset_id and job.id < offset_id:
+                continue
+
+        # Build up a buffer of jobs with the same create time. Once we get a
+        # job with an older create time we yield all the jobs in the buffer
+        # sorted by job-id + task-id
+        job_buffer.append(job)
+        if job.submission != last_create_time:
+            for j in sorted(job_buffer, key=lambda j: j.id):
+                yield j
+            job_buffer = []
+        last_create_time = job.submission
+
+    # If we hit the end of the dstat job generator, ensure to yield the jobs
+    # stored in the buffer before returning
+    for j in sorted(job_buffer, key=lambda j: j.id):
+        yield j
+
+
+def generate_jobs_count(provider,
+                        project_id,
+                        window_min,
+                        window_max=datetime.datetime.now(tz=tzlocal())):
+    """Get the generator of jobs for aggregation.
+
+            Args:
+                provider (string): type of provider
+                project_id (string): the project id
+                window_min (datetime): the earliest time of aggregation time window
+                window_max (datatime): the latest time of aggregation time window
+
+            Returns:
+                generator: Retrieved jobs
+    """
+    create_time_min = window_min - datetime.timedelta(
+        days=_MAX_PENDING_TIME_DAYS)
+
+    jobs = execute_redirect_stdout(lambda: dstat.lookup_job_tasks(
+        provider=provider,
+        statuses=None,
+        user_ids=None,
+        job_ids=None,
+        task_ids=None,
+        create_time_min=create_time_min,
+        create_time_max=window_max,
+        job_names=None,
+        labels=None))
+
+    jobs_buffer = []
+    for j in jobs:
+        job = _query_jobs_result(j, project_id)
+        # Filter jobs that do no end within the time window
+        if job.end and (job.end < window_min or job.end > window_max):
+            continue
+
+        jobs_buffer.append(job)
+
+    for job in jobs_buffer:
+        yield job
+
+
+def auth_token():
```

Comment:
I would move auth_token and provider_type back into the controller, they aren't really related to the generator.

Reply:
The `aggregation_controller` can reuse these two functions. Either we put these two functions to other `utils` file where makes better sense or move the function calls of these two functions within the generator. I will probably go with the later one.

Reply:
Both `aggragation_controller` and `job_controller` use these two functions so the second approach mentioned above is not viable. Do you have any suggestion where we should put them?

Reply:
How about `utils/providers.py`?